metashare/consumers.py

```diff
@@ -32,17 +33,42 @@ class PushNotificationConsumer(AsyncJsonWebsocketConsumer):
 
             channel_layer.group_send(group_name, {
                 'type': 'notify',  # This routes it to this handler.
-                'content': json_message,
+                'content': {
+                    'type': str (frontend Redux event),
+                    'payload': {
+                        'originating_user_id': str,
+                        'message': str (error message, optional),
+                    },
+                    'model_name': str,
+                    'id': str,
+                },
             })
         """
         # Take lock out of redis for this message:
         await clear_message_semaphore(self.channel_layer, event)
         if "content" in event:
-            await self.send_json(event["content"])
+            message = await self.hydrate_message(event["content"])
+            await self.send_json(message)
             return
 
+    async def hydrate_message(self, content):
+        content = deepcopy(content)
+        model_name = content.pop("model_name")
+        id = content.pop("id")
+        instance = await self.get_instance(model=model_name, id=id)
+        # We specifically don't want to include every user, as that
+        # would cause every error to include the user who's getting the
+        # error to be included. It'd just be noise on the wire.
+        if model_name.lower() != "user":
+            content["payload"]["model"] = instance.get_serialized_representation(
+                self.scope["user"]
+            )
+        return content
+
     @database_sync_to_async
     def get_instance(self, *, model, id, **kwargs):
+        # XXX: We currently hard-code API as it's our only
+        # model-containing app:
         Model = apps.get_model("api", model)
         return Model.objects.get(pk=id)
```

Comment:
@wlonk When there's an error creating a new org (and, I presume, other `notify_scratch_org_error` cases where the org is deleted on-error: `finalize_provision`, `remove_scratch_org`, and `finalize_refresh_org`), this throws an error:

```
id=no-job-id at=INFO time="2020-03-24 21:00:13.772743" tag=external module=worker msg="default: metashare.api.jobs.create_branches_on_github_then_create_scratch_org(originating_user_id='n10YoRw', scratch_org=<ScratchOrg: ScratchOrg object (O1Xpe1Y)>) (3244fdca-120f-45b4-9eed-55f883fbd8ca)"
id=b3214bb2f22e4f49b826644f81ff6950 at=INFO time="2020-03-24 21:00:13.792282" tag=external module=logging_middleware method=POST path=/api/scratch-orgs/ status=201 source_ip=127.0.0.1 user_agent='Mozilla/5.0 (Macintosh;=True Intel=True Mac=True OS=True X=True 10.15;=True rv:74.0)=True Gecko/20100101=True Firefox/74.0'=True time=0.6483473777770996 forwardedfor=None user=n10YoRw
HTTP POST /api/scratch-orgs/ 201 [0.66, 127.0.0.1:60426]
Traceback (most recent call last):
  File "/app/metashare/api/jobs.py", line 221, in create_branches_on_github_then_create_scratch_org
    _create_org_and_run_flow(
  File "/app/metashare/api/jobs.py", line 154, in _create_org_and_run_flow
    scratch_org_config, cci, org_config = create_org(
  File "/app/metashare/api/sf_run_flow.py", line 257, in create_org
    org_result = get_org_result(
  File "/app/metashare/api/sf_run_flow.py", line 164, in get_org_result
    response = devhub_api.ScratchOrgInfo.create(create_args)
  File "/usr/local/lib/python3.8/site-packages/simple_salesforce/api.py", line 688, in create
    result = self._call_salesforce(
  File "/usr/local/lib/python3.8/site-packages/simple_salesforce/api.py", line 818, in _call_salesforce
    exception_handler(result, self.name)
  File "/usr/local/lib/python3.8/site-packages/simple_salesforce/util.py", line 61, in exception_handler
    raise exc_cls(result.url, result.status_code, name, response_content)
simple_salesforce.exceptions.SalesforceMalformedRequest: Malformed request https://na85.salesforce.com/services/data/v47.0/sobjects/ScratchOrgInfo/. Response content: [{'message': 'The signup request failed because this organization has reached its active scratch org limit', 'errorCode': 'LIMIT_EXCEEDED', 'fields': []}]

id=no-job-id at=ERROR time="2020-03-24 21:00:17.253401" tag=external module=worker msg="simple_salesforce.exceptions.SalesforceMalformedRequest: Malformed request https://na85.salesforce.com/services/data/v47.0/sobjects/ScratchOrgInfo/. Response content: [{'message': 'The signup request failed because this organization has reached its active scratch org limit', 'errorCode': 'LIMIT_EXCEEDED', 'fields': []}]
Traceback (most recent call last):
  File \"/usr/local/lib/python3.8/site-packages/rq/worker.py\", line 884, in perform_job
    rv = job.perform()
  File \"/usr/local/lib/python3.8/site-packages/rq/job.py\", line 664, in perform
    self._result = self._execute()
  File \"/usr/local/lib/python3.8/site-packages/rq/job.py\", line 670, in _execute
    return self.func(*self.args, **self.kwargs)
  File \"/app/metashare/api/jobs.py\", line 221, in create_branches_on_github_then_create_scratch_org
    _create_org_and_run_flow(
  File \"/app/metashare/api/jobs.py\", line 154, in _create_org_and_run_flow
    scratch_org_config, cci, org_config = create_org(
  File \"/app/metashare/api/sf_run_flow.py\", line 257, in create_org
    org_result = get_org_result(
  File \"/app/metashare/api/sf_run_flow.py\", line 164, in get_org_result
    response = devhub_api.ScratchOrgInfo.create(create_args)
  File \"/usr/local/lib/python3.8/site-packages/simple_salesforce/api.py\", line 688, in create
    result = self._call_salesforce(
  File \"/usr/local/lib/python3.8/site-packages/simple_salesforce/api.py\", line 818, in _call_salesforce
    exception_handler(result, self.name)
  File \"/usr/local/lib/python3.8/site-packages/simple_salesforce/util.py\", line 61, in exception_handler
    raise exc_cls(result.url, result.status_code, name, response_content)
simple_salesforce.exceptions.SalesforceMalformedRequest: Malformed request https://na85.salesforce.com/services/data/v47.0/sobjects/ScratchOrgInfo/. Response content: [{'message': 'The signup request failed because this organization has reached its active scratch org limit', 'errorCode': 'LIMIT_EXCEEDED', 'fields': []}]
"
Exception inside application: ScratchOrg matching query does not exist.
  File "/usr/local/lib/python3.8/site-packages/channels/sessions.py", line 183, in __call__
    return await self.inner(receive, self.send)
  File "/usr/local/lib/python3.8/site-packages/channels/middleware.py", line 41, in coroutine_call
    await inner_instance(receive, send)
  File "/usr/local/lib/python3.8/site-packages/channels/consumer.py", line 58, in __call__
    await await_many_dispatch(
  File "/usr/local/lib/python3.8/site-packages/channels/utils.py", line 51, in await_many_dispatch
    await dispatch(result)
  File "/usr/local/lib/python3.8/site-packages/channels/consumer.py", line 73, in dispatch
    await handler(message)
  File "/app/metashare/consumers.py", line 50, in notify
    message = await self.hydrate_message(event["content"])
  File "/app/metashare/consumers.py", line 58, in hydrate_message
    instance = await self.get_instance(model=model_name, id=id)
  File "/usr/local/lib/python3.8/site-packages/asgiref/sync.py", line 244, in __call__
    return await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/channels/db.py", line 14, in thread_handler
    return super().thread_handler(loop, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/asgiref/sync.py", line 277, in thread_handler
    return func(*args, **kwargs)
  File "/app/metashare/consumers.py", line 73, in get_instance
    return Model.objects.get(pk=id)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 415, in get
    raise self.model.DoesNotExist(
  ScratchOrg matching query does not exist.
```

Reply:
@wlonk Also, this _could_ be a good time to bring those scratch-org notification methods in line with the pattern for other models?

Reply:
I think I addressed the deleted model situation, but let me know if it works. I'll look at normalizing the scratch-org notification methods.